### PR TITLE
Better Windows Support

### PR DESCRIPTION
**Pull request by themperek**
- Set `PYTHONHOME` in order to properly populate `sys.path` in embedded python interpreter in `Makefile.pylib.Msys`. (No need to set all pythonpath in the command line , just additional).
- Set `PATH` to include a path to cocotb compiled libraries in `Makefile.pylib.Msys`.
- Set appropriate path format for sources and paths.
- Remove unnecessary logic from example Makefiles.
- Reorder necessary variables and cleanup.

The important changes are in `cocotb/share/makefiles/Makefile.inc` and 
 `cocotb/share/makefiles/Makefile.pylib.Msys` the rest is just a cleanup (mostly remove a lot of obsolete code).

Resolves: #1064 #428 #498 #436 #354 

## Usage:
- (For Aldec) You may have to **not install** C++ compiler together with Riviera or remove/rename `mingw` directory in the install directory.
- Install [Miniconda](https://conda.io/en/latest/miniconda.html)  (according to your simulation version 32/64bit)
- Start `Anaconda Prompt` 
-  Install msys `conda install -c msys2 m2-base m2-make libpython m2w64-toolchain` 
- Set proper `%path%` to the binary of the simulator

For msys2 install [msys2](https://www.msys2.org). Install `pacman -S --needed base-devel mingw-w64-i686-python3-pip` or `pacman -S --needed mingw-w64-x86_64-toolchain mingw-w64-x86_64-python3-pip`. Use mingw32/64.exe.

## Test matrix on Windows 10 (64):

|            | icarus 10.1,64 | questa,64 | questa,32 | Rivera 2019.04,64 | Rivera 2019.04,32 | Modelsim 10.1b,32 |
|------------|----------------|----------------|----------------|-------------------|-------------------|-------------------|
| conda(3.7) | ok             | ok             | ok             | ok*              | ok*               | ok                |
| msys(3.7)   | ok             | ok             | ok             | ok               | ok                | ok                |

If this is acceptable will update documentation.
[*] Simulation is running ok but on exit `segmentation fault`.

PS. Thanks to [Aldec](https://www.aldec.com/) for providing the license.

### CODE REVIEW

File: cocotb/share/makefiles/simulators/Makefile.icarus [hunk 1]
```diff
@@ -62,21 +60,10 @@ $(SIM_BUILD)/sim.vvp: $(SIM_BUILD) $(VERILOG_SOURCES) $(CUSTOM_COMPILE_DEPS)
...
-else
-LIB_LOAD := LD_LIBRARY_PATH=$(LIB_DIR):$(LD_LIBRARY_PATH)
-NEW_PYTHONPATH := $(PYTHONPATH)
-endif
-
 $(COCOTB_RESULTS_FILE): $(SIM_BUILD)/sim.vvp $(CUSTOM_SIM_DEPS) $(COCOTB_LIBS) $(COCOTB_VPI_LIB)
 	-@rm -f $(COCOTB_RESULTS_FILE)
 
-	PYTHONPATH=$(LIB_DIR):$(PWD):$(NEW_PYTHONPATH) $(LIB_LOAD) MODULE=$(MODULE) \
+	PYTHONPATH=$(PYTHONPATH) MODULE=$(MODULE) \
```
Comment (garmin-mjames) [inline cocotb/share/makefiles/simulators/Makefile.icarus]: I'm not sure why, but with Windows 10 WSL (Ubuntu 18.04 LTS) removing `LD_LIBRARY_PATH` causes `iverilog` to be unable to find some of the libraries, causing a `dlopen()` error.

Comment (themperek) [reply cocotb/share/makefiles/simulators/Makefile.icarus]: This is very strange configuration. It looks like Windows WSL does not respect `Library runpath` 
One can go around this bay setting:  `LD_LIBRARY_PATH=$(cocotb-config --prefix)/cocotb/libs`

Comment (garmin-mjames) [reply cocotb/share/makefiles/simulators/Makefile.icarus]: Once this is in, it'd be nice to document that workaround somewhere (on a wiki page?)

Comment (themperek) [reply cocotb/share/makefiles/simulators/Makefile.icarus]: did you tried WSL 2?

File: cocotb/share/def/modelsim.def [hunk 2]
```diff
@@ -100,3 +100,16 @@ mti_TickLeft
...
+Tcl_ResetResult
+Tcl_GetObjResult
+Tcl_ResetResult
+Tcl_ListObjGetElements
+Tcl_GetStringResult
+TclFreeObj
+Tcl_ResetResult
+Tcl_ResetResult
+Tcl_GetString
+TclFreeObj
```
Comment (eric-wieser) [inline cocotb/share/def/modelsim.def:115]: Without these lines, modelsim is unusable on windows, right? Or is that true of this whole PR anyway?

Comment (themperek) [reply cocotb/share/def/modelsim.def:115]: Without those lines, it will not install. 

It Will not stimulate without the rest.

Comment (themperek) [reply cocotb/share/def/modelsim.def:115]: To be clear will not install with FLI support.

File: cocotb/share/makefiles/Makefile.pylib.Msys [hunk 3]
```diff
@@ -47,3 +47,20 @@ ifeq ($(PYTHON_LIBDIR),None)
...
+endif
+ifneq ($(VHDL_SOURCES),)
+    VHDL_SOURCES := $(shell cygpath -m $(VHDL_SOURCES))
+endif
+
+# On Windows there is no support for @rpath so we need to set PATH to find cocotb libraries
+export PATH := $(LIB_DIR):$(PATH)
+
+# Set PYTHONHOME to properly populate sys.path in embedded python interpreter
+export PYTHONHOME := $(shell $(PYTHON_BIN) -c "from distutils.sysconfig import get_config_var; print(get_config_var('prefix'))")
```
Comment (eric-wieser) [inline cocotb/share/makefiles/Makefile.pylib.Msys:63]: I think we want to do one of two things:

* Eliminate `PYTHON_BIN`, and use something from `cocotb-config` instead
* Eliminate `cocotb-config`, and use `$(PYTHON_BIN) -m cocotb.config` instead

Otherwise, we can end up mixing two different python installations.

Comment (themperek) [reply cocotb/share/makefiles/Makefile.pylib.Msys:63]: Yes. It was not available at the time. I can rebase and clean this up a bit now. I like `Middle-ground`.

Comment (eric-wieser) [reply cocotb/share/makefiles/Makefile.pylib.Msys:63]: In the long run, I'm wondering if eliminating `cocotb-config` is the way to go.

Comment (themperek) [reply cocotb/share/makefiles/Makefile.pylib.Msys:63]: Long run eliminating Makefiles maybe ;-)  ?

File: cocotb/share/makefiles/Makefile.pylib.Msys [hunk 4]
```diff
@@ -47,3 +47,20 @@ ifeq ($(PYTHON_LIBDIR),None)
...
+# For tcl scripts purpose source files paths in form of C:/User/..
+ifneq ($(VERILOG_SOURCES),)
+    VERILOG_SOURCES := $(shell cygpath -m $(VERILOG_SOURCES))
+endif
+ifneq ($(VHDL_SOURCES),)
+    VHDL_SOURCES := $(shell cygpath -m $(VHDL_SOURCES))
+endif
+
+# On Windows there is no support for @rpath so we need to set PATH to find cocotb libraries
+export PATH := $(LIB_DIR):$(PATH)
```
Comment (imphil) [inline cocotb/share/makefiles/Makefile.pylib.Msys:60]: PATH is used for libraries on Windows? If that's the case it needs a longer comment please.

File: cocotb/share/makefiles/Makefile.pylib.Msys (hunk 3 above)
Comment (imphil) [reply cocotb/share/makefiles/Makefile.pylib.Msys:63]: This line is about PYTHONHOME, not PYTHON_BIN. PYTHON_BIN can already be set from `cocotb-config --python-bin`, we might want to to the same for PYTHONHOME, or keep the line of Python as it is in this patch.

File: cocotb/share/makefiles/simulators/Makefile.aldec [hunk 5]
```diff
@@ -144,35 +137,19 @@ ifeq ($(COVERAGE),1)
...
 
 	# check that the file was actually created
 	test -f $(COCOTB_RESULTS_FILE)
 
 clean::
 	@rm -rf $(SIM_BUILD)
-
+	@rm -rf compile
+	@rm -rf library.cfg
+	@rm -rf dataset.asdb
```
Comment (imphil) [inline cocotb/share/makefiles/simulators/Makefile.aldec:155]: That's an artifact of removing the `cd $(SIM_BUILD)` above? Can you comment why this change of behavior is necessary or better?

File: cocotb/share/makefiles/simulators/Makefile.icarus (hunk 1 above)
Comment (imphil) [reply cocotb/share/makefiles/simulators/Makefile.icarus:66]: WSL1 is dead, let's add a doc/wiki entry somewhere (probably sufficient to say "it's not supported").

File: cocotb/_build_libs.py [hunk 6]
```diff
@@ -59,6 +59,9 @@ def get_ext_filename(self, ext_name):
         head, tail = os.path.split(filename)
         tail_split = tail.split(".")
 
+        # mingw on msys2 uses `-` as seperator
+        tail_split = tail_split[0].split("-")
```
Comment (imphil) [inline cocotb/_build_libs.py:63]: Please adjust the function comment to mention this behavior.

File: cocotb/share/makefiles/Makefile.pylib.Msys [hunk 7]
```diff
@@ -47,3 +47,20 @@ ifeq ($(PYTHON_LIBDIR),None)
 endif
 
 PYTHON_INCLUDEDIR := $(shell $(PYTHON_BIN) -c 'import distutils.sysconfig; print( distutils.sysconfig.get_python_inc() )')
+
+# For tcl scripts purpose source files paths in form of C:/User/..
```
Comment (imphil) [inline cocotb/share/makefiles/Makefile.pylib.Msys:51]: VERILOG_SOURCES and the other variables in this files are not only used by TCL script; they are sometimes directly passed to simulators, and used as dependency in Makefiles. Is it just the comment that's wrong, and msys paths should (or can) always be Windows-style paths with forward slashes?  Can you adjust the comment accordingly?

**Comment by themperek** [comment]
I think I will simplify this even more and split a few smaller PRs. If this OK for you?

File: cocotb/share/makefiles/Makefile.pylib.Msys (hunk 3 above)
Comment (themperek) [reply cocotb/share/makefiles/Makefile.pylib.Msys:63]: I will prepare separate PR.

File: cocotb/share/makefiles/simulators/Makefile.aldec (hunk 5 above)
Comment (themperek) [reply cocotb/share/makefiles/simulators/Makefile.aldec:155]: To be consistent with this: https://github.com/cocotb/cocotb/pull/1063/files all those files are created in `pwd` directory (same for modelsim).

**Comment by themperek** [comment]
Closing in favor of #1588 and potentially #1565.

File: cocotb/share/makefiles/simulators/Makefile.icarus (hunk 1 above)
Comment (themperek) [reply cocotb/share/makefiles/simulators/Makefile.icarus:66]: @garmin-mjames I know what is wrong will try to make PR tomorrow. It is a more general issue.

Comment (themperek) [reply cocotb/share/makefiles/simulators/Makefile.icarus:66]: This should fix this https://github.com/cocotb/cocotb/pull/1606
